src/rust/runtime/memory/buffer/demibuffer.rs

```diff
@@ -0,0 +1,820 @@
+    // the behavior of DPDK's rte_pktmbuf_trim() routine.
+    pub fn trim(&mut self, nbytes: u16) -> Result<(), Fail> {
+        // ToDo: Review having this "match", since MetaData and MBuf are laid out the same, these are equivalent cases.
+        match self.get_tag() {
+            TAG_HEAP => {
+                let md_first: &mut MetaData = self.get_metadata();
+                let md_last: &mut MetaData = md_first.get_last_segment();
+
+                if nbytes > md_last.data_len {
+                    return Err(Fail::new(libc::EINVAL, "tried to remove more bytes than are present"));
+                }
+                md_last.data_len -= nbytes;
+                md_first.pkt_len -= nbytes as u32;
+            },
+            #[cfg(feature = "libdpdk")]
+            TAG_DPDK => {
+                // Safety: rte_pktmbuf_trim is a FFI, which is safe since we call it with an actual MBuf pointer.
+                if unsafe { rte_pktmbuf_trim(self.get_mbuf(), nbytes) } != 0 {
+                    return Err(Fail::new(libc::EINVAL, "tried to remove more bytes than are present"));
+                }
+            },
+            _ => {
+                panic!("corrupted DemiBuffer");
+            },
+        }
+
+        Ok(())
+    }
+
+    // Splits off a new DemiBuffer containing a subset of the data in this DemiBuffer, starting at the given offset.
+    // The data contained in the new DemiBuffer is removed from the original DemiBuffer.
+    // Note: the DemiBuffer being split must be a single buffer segment (not a chain) large enough to hold `offset`.
+    pub fn split(&mut self, offset: u16) -> Result<Self, Fail> {
+        // Check that a split is allowed.
+        match self.get_tag() {
+            TAG_HEAP => {
+                let md_front: &mut MetaData = self.get_metadata();
+                if md_front.nb_segs != 1 {
+                    return Err(Fail::new(libc::EINVAL, "attempted to split multi-segment DemiBuffer"));
+                }
+                if md_front.data_len < offset {
+                    return Err(Fail::new(libc::EINVAL, "split offset is more bytes than are present"));
+                }
+            },
+            #[cfg(feature = "libdpdk")]
+            TAG_DPDK => {
+                let mbuf: *mut rte_mbuf = self.get_mbuf();
+                // Safety: The `mbuf` dereferences in this block are safe, as it is aligned and dereferenceable.
+                unsafe {
+                    if (*mbuf).nb_segs != 1 {
+                        return Err(Fail::new(libc::EINVAL, "attempted to split multi-segment DemiBuffer"));
+                    }
+                    if (*mbuf).data_len < offset {
+                        return Err(Fail::new(libc::EINVAL, "split offset is more bytes than are present"));
+                    }
+                }
+            },
+            _ => {
+                panic!("corrupted DemiBuffer");
+            },
```

Comment:
Please, log and fail. Someone may simply discard this buffer and continue.

Reply:
These are "cannot happen" (outside of hardware failure) situations that are placed here because the compiler insists upon having a "catch-all" case for the match operator.  If memory reads are returning bogus values, then panicking is appropriate.

Reply:
- If you were using the enumeration, that would be not possible.
- To make the compiler happy about non-covered cases on conditional compilation use `#[allow(unreachable_patterns)]` and change `panic()` to `unreachable()`

Reply:
Switching tag values from consts to an enum removed the need for "catch-all" cases on those match operations, so this is now moot.